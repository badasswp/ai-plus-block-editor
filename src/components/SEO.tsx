import { __ } from '@wordpress/i18n';
import { check } from '@wordpress/icons';
import { useState, useEffect } from '@wordpress/element';
import { store as noticesStore } from '@wordpress/notices';
import { Button, TextareaControl, Icon } from '@wordpress/components';
import { useSelect, useDispatch } from '@wordpress/data';
import apiFetch from '@wordpress/api-fetch';

import Toast from '../components/Toast';

interface selectProps {
	getCurrentPostId: () => number;
	getEditedPostContent: () => any;
	getEditedPostAttribute: ( attribute: string ) => any;
}

/**
 * SEO.
 *
 * This Component returns the SEO keywords,
 * label and button.
 *
 * @since 1.1.0
 *
 * @return {JSX.Element} SEO Keywords.
 */
const SEO = (): JSX.Element => {
	const editorStore = 'core/editor';
	const [ keywords, setKeywords ] = useState( '' );
	const [ isLoading, setIsLoading ] = useState( false );
	const { editPost, savePost } = useDispatch( editorStore ) as any;
	const { createErrorNotice, removeNotice } = useDispatch( noticesStore );
	const { postId, postContent, postKeywords, notices } = useSelect(
		( select ) => {
			const { getNotices } = select( noticesStore );
			const {
				getCurrentPostId,
				getEditedPostContent,
				getEditedPostAttribute,
			} = select( editorStore ) as selectProps;

			return {
				postId: getCurrentPostId(),
				postContent: getEditedPostContent(),
				postKeywords:
					getEditedPostAttribute( 'meta' )?.apbe_seo_keywords,
				notices: getNotices(),
			};
		},
		[]
	);

	useEffect( () => {
		setKeywords( postKeywords );
	}, [ postKeywords ] );

	/**
	 * This function fires when the user clicks
	 * the `Generate` button.
	 *
	 * @since 1.1.0
	 *
	 * @return { void }
	 */
	const handleClick = async (): Promise< void > => {
		notices.forEach( ( notice ) => removeNotice( notice.id ) );
		setIsLoading( true );

		try {
			const aiKeywords: string = await apiFetch( {
				path: '/ai-plus-block-editor/v1/sidebar',
				method: 'POST',
				data: {
					id: postId,
					text: postContent?.text || postContent,
					feature: 'keywords',
				},
			} );

			/**
			 * This function returns a promise that resolves
			 * to the AI generated SEO keywords when the Animation
			 * responsible for showing same is completed.
			 *
			 * @since 1.2.0
			 *
			 * @return { Promise<string> } Animated text.
			 */
			const showAnimatedAiText = (): Promise< string > => {
				let limit = 1;

				return new Promise( ( resolve ) => {
					const animatedTextInterval = setInterval( () => {
						if ( aiKeywords.length === limit ) {
							clearInterval( animatedTextInterval );
							resolve( aiKeywords );
						}
						setKeywords( aiKeywords.substring( 0, limit ) );
						limit++;
					}, 5 );
				} );
			};

			showAnimatedAiText().then( ( newKeywords ) => {
				editPost( { meta: { apbe_seo_keywords: newKeywords } } );
			} );

			setIsLoading( false );
		} catch ( e ) {
			createErrorNotice( e.message );
			setIsLoading( false );
		}
	};

	/**
	 * This function fires when the user selects
	 * the AI generated result.
	 *
	 * @since 1.1.0
	 *
	 * @return { void }
	 */
	const handleSelection = (): void => {
		editPost( { meta: { apbe_seo_keywords: keywords } } );
		savePost();
	};

	return (
		<>
			<p>
				<strong>
					{ __( 'SEO Keywords', 'ai-plus-block-editor' ) }
				</strong>
			</p>
			<TextareaControl
				rows={ 7 }
				value={ keywords }
				onChange={ ( text ) => setKeywords( text ) }
				__nextHasNoMarginBottom
			/>
			<div className="apbe-button-group">
				<Button variant="primary" onClick={ handleClick }>
					{ __( 'Generate', 'ai-plus-block-editor' ) }
				</Button>
				<Button variant="secondary" onClick={ handleSelection }>
					<Icon icon={ check } />
				</Button>
			</div>
			<Toast
				message={ __(
					'AI is generating text, please hold on for a bitâ€¦'
				) }
				isLoading={ isLoading }
			/>
		</>
	);
};

export default SEO;
