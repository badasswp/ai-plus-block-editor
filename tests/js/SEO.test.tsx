import '@testing-library/jest-dom';
import { act, render, fireEvent, waitFor } from '@testing-library/react';

import apiFetch from '@wordpress/api-fetch';
import { useSelect, useDispatch } from '@wordpress/data';

import SEO from '../../src/components/SEO';

jest.mock( '@wordpress/i18n', () => ( {
	__: jest.fn( ( arg ) => arg ),
} ) );

jest.mock( '@wordpress/data', () => ( {
	useSelect: jest.fn(),
	useDispatch: jest.fn(),
} ) );

jest.mock( '@wordpress/notices', () => ( {
	store: 'core/notices',
} ) );

jest.mock( '@wordpress/components', () => ( {
	Button: jest.fn( ( { children, variant, onClick } ) => {
		return (
			<>
				<button className={ variant } onClick={ onClick }>
					{ children }
				</button>
			</>
		);
	} ),

	Icon: jest.fn( () => {
		return <>Icon</>;
	} ),

	TextareaControl: jest.fn( ( { rows, value, onChange } ) => {
		return (
			<>
				<textarea
					rows={ rows }
					onChange={ ( e ) => onChange( e.target.value ) }
					value={ value }
				/>
			</>
		);
	} ),
} ) );

jest.mock( '@wordpress/api-fetch', () => jest.fn() );

describe( 'SEO', () => {
	beforeEach( () => {
		( useSelect as jest.Mock ).mockReturnValue( {
			postId: 1,
			postContent: 'Hello World',
			postKeywords: 'AI generated SEO...',
			notices: [],
		} );

		( useDispatch as jest.Mock ).mockReturnValue( {
			editPost: jest.fn(),
			savePost: jest.fn(),
			removeNotice: jest.fn(),
			createErrorNotice: jest.fn(),
		} );

		jest.spyOn( console, 'error' ).mockImplementation( () => {} );
	} );

	afterEach( () => {
		jest.clearAllMocks();
	} );

	it( 'renders SEO Keywords component', () => {
		const { container, getByText, getByRole } = render( <SEO /> );

		expect( container ).toMatchSnapshot();

		expect( getByText( 'AI generated SEO...' ) ).toBeInTheDocument();
		expect( getByText( 'SEO Keywords' ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'Icon' } ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'Icon' } ) ).toHaveClass(
			'secondary'
		);
		expect( getByRole( 'button', { name: 'Generate' } ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'Generate' } ) ).toHaveClass(
			'primary'
		);
	} );

	it( 'renders fetched API data from AI LLM', async () => {
		const mockEditPost = jest.fn();
		( useDispatch as jest.Mock ).mockReturnValue( {
			editPost: mockEditPost,
		} );

		( apiFetch as unknown as jest.Mock ).mockImplementation(
			jest.fn( () => Promise.resolve( 'hello, world, beautiful' ) )
		);

		const { getByText, getByRole } = render( <SEO /> );

		expect( getByText( 'AI generated SEO...' ) ).toBeInTheDocument();

		const button = getByRole( 'button', { name: 'Generate' } );
		await act( async () => {
			fireEvent.click( button );
		} );

		await waitFor( () => {
			expect( mockEditPost ).toHaveBeenCalledTimes( 1 );
			expect(
				getByText( 'hello, world, beautiful' )
			).toBeInTheDocument();
		} );
	} );

	it( 'renders error notice on API fail', async () => {
		const mockCreateErrorNotice = jest.fn();
		( useDispatch as jest.Mock ).mockReturnValue( {
			createErrorNotice: mockCreateErrorNotice,
		} );

		( apiFetch as unknown as jest.Mock ).mockRejectedValueOnce(
			new Error( 'AI LLM down...' )
		);

		const { getByText, getByRole } = render( <SEO /> );

		expect( getByText( 'AI generated SEO...' ) ).toBeInTheDocument();

		const button = getByRole( 'button', { name: 'Generate' } );
		await act( async () => {
			fireEvent.click( button );
		} );

		await waitFor( () => {
			expect( mockCreateErrorNotice ).toHaveBeenCalledTimes( 1 );
			expect( getByText( 'AI generated SEO...' ) ).toBeInTheDocument();
		} );
	} );

	it( 'saves the selected AI SEO Keywords', async () => {
		const mockEditPost = jest.fn();
		const mockSavePost = jest.fn();
		( useDispatch as jest.Mock ).mockReturnValue( {
			editPost: mockEditPost,
			savePost: mockSavePost,
		} );

		( apiFetch as unknown as jest.Mock ).mockImplementation(
			jest.fn( () => Promise.resolve( 'hello, world, beautiful' ) )
		);

		const { getByText, getByRole } = render( <SEO /> );

		expect( getByText( 'AI generated SEO...' ) ).toBeInTheDocument();

		const button = getByRole( 'button', { name: 'Generate' } );
		await act( async () => {
			fireEvent.click( button );
		} );

		await waitFor( () => {
			expect( mockEditPost ).toHaveBeenCalledTimes( 1 );
			expect(
				getByText( 'hello, world, beautiful' )
			).toBeInTheDocument();
		} );

		const icon = getByRole( 'button', { name: 'Icon' } );
		await act( async () => {
			fireEvent.click( icon );
		} );

		await waitFor( () => {
			expect( mockEditPost ).toHaveBeenCalledTimes( 2 );
			expect( mockSavePost ).toHaveBeenCalledTimes( 1 );
		} );
	} );
} );
