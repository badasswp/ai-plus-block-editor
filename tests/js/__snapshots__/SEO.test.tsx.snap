// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEO renders SEO Keywords component 1`] = `
<div>
  <p>
    <strong>
      SEO Keywords
    </strong>
  </p>
  <textarea
    rows="7"
  >
    AI generated SEO...
  </textarea>
  <div
    class="apbe-button-group"
  >
    <button
      class="primary"
    >
      Generate
    </button>
    <button
      class="secondary"
    >
      Icon
    </button>
  </div>
</div>
`;
